lxd/db/cluster/entities.go

```diff
@@ -733,3 +733,237 @@ func GetEntityURLs(ctx context.Context, tx *sql.Tx, projectName string, entityTy
+
+// groupIDFromURL gets the ID of a group from its URL.
+var groupIDFromURL = `SELECT ?, groups.id FROM groups WHERE '' = ? AND '' = ? AND groups.name = ?`
+
+// identityProviderGroupIDFromURL gets the ID of a identityProviderGroup from its URL.
+var identityProviderGroupIDFromURL = `SELECT ?, identity_provider_groups.id FROM identity_provider_groups WHERE '' = ? AND '' = ? AND identity_provider_groups.name = ?`
+
+// identityIDFromURL gets the ID of a identity from its URL.
+var identityIDFromURL = fmt.Sprintf(`SELECT ?, identities.id FROM identities WHERE '' = ? AND '' = ? AND CASE identities.auth_method WHEN %d THEN '%s' WHEN %d THEN '%s' END = ? and identities.identifier = ?`, authMethodTLS, api.AuthenticationMethodTLS, authMethodOIDC, api.AuthenticationMethodOIDC)
+
+// identityIDFromURLStatements is a map of entity.Type to a statement that can be used to get the ID of the entity from its URL.
+var entityIDFromURLStatements = map[entity.Type]string{
+	entity.TypeContainer:             containerIDFromURL,
+	entity.TypeImage:                 imageIDFromURL,
+	entity.TypeProfile:               profileIDFromURL,
+	entity.TypeProject:               projectIDFromURL,
+	entity.TypeCertificate:           certificateIDFromURL,
+	entity.TypeInstance:              instanceIDFromURL,
+	entity.TypeInstanceBackup:        instanceBackupIDFromURL,
+	entity.TypeInstanceSnapshot:      instanceSnapshotIDFromURL,
+	entity.TypeNetwork:               networkIDFromURL,
+	entity.TypeNetworkACL:            networkACLIDFromURL,
+	entity.TypeNode:                  nodeIDFromURL,
+	entity.TypeOperation:             operationIDFromURL,
+	entity.TypeStoragePool:           storagePoolIDFromURL,
+	entity.TypeStorageVolume:         storageVolumeIDFromURL,
+	entity.TypeStorageVolumeBackup:   storageVolumeBackupIDFromURL,
+	entity.TypeStorageVolumeSnapshot: storageVolumeSnapshotIDFromURL,
+	entity.TypeWarning:               warningIDFromURL,
+	entity.TypeClusterGroup:          clusterGroupIDFromURL,
+	entity.TypeStorageBucket:         storageBucketIDFromURL,
+	entity.TypeImageAlias:            imageAliasIDFromURL,
+	entity.TypeNetworkZone:           networkZoneIDFromURL,
+	entity.TypeGroup:                 groupIDFromURL,
+	entity.TypeIdentityProviderGroup: identityProviderGroupIDFromURL,
+	entity.TypeIdentity:              identityIDFromURL,
+}
+
+// GetEntityReferencesFromURLs accepts a slice of *api.URL and returns a map of *api.URL to *EntityRef. The pointers in
+// input slice are directly used as keys in the output map. This function will return an error if any of the input URLs
+// do not resolve to an ID (e.g. if they are not valid, or could not be found).
+func GetEntityReferencesFromURLs(ctx context.Context, tx *sql.Tx, entityURLs []*api.URL) (map[*api.URL]*EntityRef, error) {
```

Comment:
Is there a reason you're accepting a slice of URL pointers rather than just URls, same for the return values. Why a map of URL pointers with pointer EntityRef values? Seems odd given that maps are returned by reference anyway.

Reply:
You've said 

> The pointers in input slice are directly used as keys in the output map.

But that hasn't told me *why* this is needed, it seems quite unorthodox.